ZAP: Encryption for mobile push notifications

prakhar1144 opened the pull request:
[Rendered](https://github.com/prakhar1144/zulip-architecture/blob/encrypt-push-notification/zaps/0002-encrypt-push-notifications.md)

prakhar1144 commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,182 @@
+
+# Detailed design
+
+## High-Level Plan
```
If we think this subsection is not useful, we can remove it. I just added it as a brief overview for someone new reading it -- instead of directly jumping into data model.

prakhar1144 commented on zaps/0002-encrypt-push-notifications.md:
````diff
@@ -0,0 +1,182 @@
+  ```
+- On the push bouncer, add a new table (parallel to `RemotePushDeviceToken`):
+  ```
+  class zilencer.models.RemotePushDevice:
````
TODO: There's a comment thread in the doc regarding "how the import tool should handle this table", I need to document it here.

prakhar1144 replied:
Done in second commit.

prakhar1144 replied:
Actually, we don't need to import `PushDevice`.

See: https://github.com/zulip/zulip-architecture/pull/2#discussion_r2063815370

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,182 @@
+- Client checks the server if it supports E2E encryption. If it doesn't,
+  use the legacy push notification registration flow.
+
+- Client makes an API call to the server to register itself for push notifications,
+  receives a shared secret key in response.
```
clarify it's shared between client and server

(There are a lot of shared secrets in this overall system — there's the push token, and there's the secret the server uses to talk to the bouncer. So good to make explicit.)

prakhar1144 replied:
```diff
 - Client makes an API call to the server to register itself for push notifications,
-  receives a shared secret key in response.
+  receives a new secret key (shared between client and server) in response.
```

prakhar1144 replied:
The above comment is outdated.

Updated the text to align with our plan to use public-key encryption.

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,182 @@
+- Client makes an API call to the server to register itself for push notifications,
+  receives a shared secret key in response.
+
+- While sending notification, the server encrypts the notification content
+  and sends it to push bouncer. Push bouncer forwards it to FCM/APNs.
```
clarify it's encrypted using the shared secret from above

prakhar1144 replied:
The above comment is outdated.

Updated the text to align with our plan to use public-key encryption.

gnprice commented on zaps/0002-encrypt-push-notifications.md:
````diff
@@ -0,0 +1,182 @@
+  ```
+  model Account:
+    …
+    push_account_id: RandomID[32]  # Generated by client; unique here.
````
This notation is ambiguous; 32 of what?

As I wrote on the earlier document, I think a random 32 bits is good here. It only needs to be unique within the client's local list of accounts.

timabbott replied:
OK, I think a 64-bit integer would be the right type for this, then, rather than doing something fancy.

gnprice replied:
Ah, here's something we didn't catch in the previous discussions:

In order to support the operation below where the /api/register-queue response lists the registered devices, and the client looks and sees if that means it needs to retry registering itself (https://github.com/zulip/zulip-architecture/pull/2#discussion_r2043172851), we need this to be unique with high probability among all this user's devices.

("High probability", because it's not enough to have the server enforce uniqueness: the point is that the `push_account_id` this client made up, and may not have managed to tell the server yet at all, shouldn't happen to be the same as the one that some other client successfully sent to the server.)

I think 32 bits is still realistically plenty. If someone buys a new phone *and* a new tablet every year, and keeps them around for a decade, that's 20 devices (a ludicrous number)… and at 32 bits they'd have a probability <5e-8 of collision.

But if 100 million Zulip users did that, some of them would experience collisions. So to bring the system back to overkill, we could make it 48 bits, or 64 bits.

gnprice replied:
(My last comment crossed with yours because GitHub remains lamentably bad at real-time discussion. Seems like we have a shared conclusion: 64 bits.)

prakhar1144 replied:
```diff
-    push_account_id: RandomID[32]  # Generated by client; unique here.
+    push_account_id: int  # 64-bit random integer generated by client; unique here.
```

timabbott replied:
Sure, that works.

gnprice commented on zaps/0002-encrypt-push-notifications.md:
````diff
@@ -0,0 +1,182 @@
+    token: str
+    timestamp: datetime
+  ```
+- On the server, add a new table (parallel to `PushDeviceToken`):
````
Does "parallel to" mean something concrete? Just "coexisting with", or is something more intended?

timabbott replied:
Coexisting with seems right. The point is that we're just making a new table, not extending the existing tables we have for the legacy unencrypted flow.

prakhar1144 replied:
Made it:
> On the server, add a new table (coexists with `PushDeviceToken`):

gnprice commented on zaps/0002-encrypt-push-notifications.md:
````diff
@@ -0,0 +1,182 @@
+    push_key: SymmetricKey  # Shared key between client and server.
+  ```
+  ```
+  // A helper data structure
+  struct PushRegistration:
````
This isn't really "on the client" — it's part of the wire protocol, and it appears in the server's PushDevice model below.

prakhar1144 replied:
"On the client" here was intended as data-structure defined on the client side. Made it an independent bullet point
````
- A helper data structure
  ```
  struct PushRegistration:
    token_kind: "apns" | "fcm"
    token: str
    timestamp: datetime
  ```
````

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,182 @@
+   **On success**, the bouncer replies to the server with `device_id`.
+   The server stores `device_id` in the `PushDevice`.
+    
+   **On error**, the server keep retrying for a long time. Like maybe
+   after a day we’d pass the error to the server admin and the user
+   (because a day is longer than any minor outage), but we’d keep retrying for a week.
```
No use retrying after the expected allowed time has elapsed, though. So the retry timescales here interact with the choice of how tight the bouncer's expectation is, above, on the liveness of `timestamp`.

prakhar1144 replied:
```diff
-   **On error**, the server keep retrying for a long time. Like maybe
-   after a day we’d pass the error to the server admin and the user
-   (because a day is longer than any minor outage), but we’d keep retrying for a week.
-    
-   Then finally after a long time perhaps we give up retrying and forget
-   that `push_account_id`. That causes the client to retry from scratch,
+   **On error**:
+   - If it's a liveness error, don't retry and forget the `push_account_id`.
+     Pass the error to the server admin and the user.
+
+   - Otherwise, the server keeps retrying for `PUSH_REGISTRATION_LIVENESS_TIMEOUT` seconds.
+
+   Forgetting the `push_account_id` causes the client to retry from scratch,
    if it’s still in use. That way if the client keeps being used, we keep
```

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,182 @@
+   The push bouncer:
+   1. Decrypts `push_registration`.
+
+   1. Checks the `timestamp` for liveness, probably 60s is a good age limit.
```
Could be, but more like a day was the design in the previous discussion a couple of years ago.

gnprice replied:
Copying from a thread there:
> [Having this liveness check] prevents replays, whether as part of an attack or due to retries in the presence of connection failures.
>
> Because we also check for duplicates, this can be a long limit; the only reason to keep it short is so that we can eventually GC records that we learn from APNs/FCM are stale or invalid, because we don’t want to do so while there could still be a valid old request to add the same record in flight.

timabbott replied:
OK, maybe 24 hours is a fine default then? We might want to name this value, say `PUSH_REGISTRATION_LIVENESS_TIMEOUT = 60 * 60 * 24`, with the intent that we'd give it a nice comment in the source code.

prakhar1144 replied:
```diff
-   1. Checks the `timestamp` for liveness, probably 60s is a good age limit.
+   1. Check the `timestamp` for liveness, `PUSH_REGISTRATION_LIVENESS_TIMEOUT`
+      which equals `24 * 60 * 60` seconds is a good age limit because
+      a day is longer than any minor outage.
+
       A coded `JsonableError` should be used if the timestamp is stale.
```

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,182 @@
+   and `encrypted_content`.
+
+1. The client uses `push_account_id` to look up the right `push_key`,
+   decrypts `encrypted_content` with it, and ends up with a set of
+   notifications to process.
+
+1. The client sorts the notifications by `message_id`, with `remove`
+   notifications cancelling `add` notifications, and passes them into
+   its display layer.
```
I don't understand why a client would end up with more than one notification at a time here.

I think this results from a comment thread on the old doc getting misread. There was a question about whether the notification-messages need a sequence number. The answer was that they already have all the information needed to order them properly: sort by message ID, and for any given message the remove goes after the add. The app already correctly does this, except that for the sorting to be fully effective it would need to keep a buffer of recent notification-messages so that a remove could cancel a later-arriving add for the same message.

In any case that buffering would be entirely orthogonal to E2EE.

timabbott replied:
Yeah, I guess the confusing was about the batching a few lines up; but that's not batching that will come through at the FCM/APNs layer -- due to the ~4K character limits on say FCM notices, we'd still be doing that API layer individually per event, more or less.

gnprice replied:
Yeah. And those batches will typically have just one entry per device anyway — each of those batches will be caused by one message getting sent, or edited, or deleted, or read, so will contain one item per device per user that's getting/got a notification.

prakhar1144 replied:
I think, this is somewhat not perfect:
```
1. The server makes a batched API call to the push bouncer (authenticated
   with the server’s API key) with arguments `device_id` and `encrypted_content`.
   - (That is, the HTTP request has a list of `(device_id, encrypted_content)` pairs.)
   - (It’s OK, just inefficient, if all batches generated by the server are,
     in practice, 1 notice in an initial version)
```

My understanding is:
* for each missed message (which `handle_push_notification` receives) we'll fetch the `bouncer_device_id`s associated with that user 
* Make an API call to bouncer with arguments `apns_encrypted_content`, `fcm_encrypted_content` & list of `bouncer_device_id`s (instead of a list of `(device_id, encrypted_content)` pairs as proposed).

We can't map `device_id` with `encrypted_content` (to create the pairs) without knowing the `token_kind`, which is encrypted for the server.

gnprice replied:
Chat thread: [#api design > E2EE push notifications @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/E2EE.20push.20notifications/near/2157498)

prakhar1144 replied:
PR is updated (3rd commit) with the decision made in that thread.

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,182 @@
+  that the `device_id` is no longer valid, and the server removes it.
+  
+- The server gives the client a list of the user’s currently valid `push_account_id`s
+  as part of `register` response, and keeps it updated through events.
```
```suggestion
  as part of the [register-queue](https://zulip.com/api/register-queue) response, and keeps it updated through events.
```

two nits:
* "the"
* please refer to API endpoints by their canonical API names :slightly_smiling_face:, i.e. our chosen OpenAPI "operation IDs" — those are the names they appear under in the API docs, and in API binding libraries such as the one we have in each mobile client

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,182 @@
+
+## Data model
+
+- On the client:
```
```suggestion
- On the client, include the following fields on [the Account type](https://github.com/zulip/zulip-flutter/blob/ec9aa351b3050a0fa3e467007826576110fdbf99/lib/model/database.dart#L66-L94):
```

prakhar1144 replied:
Done

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,182 @@
+   1. Checks the `timestamp` for liveness, probably 60s is a good age limit.
+      A coded `JsonableError` should be used if the timestamp is stale.
+
+   1. Checks that it doesn’t already have a `RemotePushDevice` for this `token`
+      and `push_account_id`. This check helps to avoid creating a duplicate
+      record if the client retries when a previous attempt actually succeeded.
```
OTOH if it does already have such a record, it returns success with that `device_id`. In particular that’s the way we expect things to be if the remote server previously made the request, didn’t get the response (due to network failure or whatever), and retried.

prakhar1144 replied:
Should we make it:
```
 Checks that it doesn’t already have a `RemotePushDevice` for this `token` and `push_account_id`. If it does 
 already have such a record, it returns success with the `device_id` corresponding to that record. This helps to 
 avoid creating a duplicate record if the server previously made the request, didn’t get the success response 
 (due to network failure or whatever), and retried.
```

The "client retries" would be handled by the remote server to check if `bouncer_device_id` already exists ? -- If it doesn't, check if `push_account_id` exists ? ( It means, the async step to register with bouncer is still WIP). If none of them exists, then registration can take place.

gnprice replied:
That update LGTM.

I don't quite follow the question after it, though. I think you're asking about what the server does when the client retries a registration request?

prakhar1144 replied:
I just meant to say --> "Server will handle things when the client retries a registration request, so no need to document it to sound like bouncer handles it (which we're doing currently)." The last para was about an overview of the various checks that the remote server can make when client-retries.

prakhar1144 replied:
Functionally no change, if that is deleted. 

It can just help with a better response message: "Already registered" vs "Registration in progress".

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,182 @@
+   **On success**, the bouncer replies to the server with `device_id`.
+   The server stores `device_id` in the `PushDevice`.
+    
+   **On error**, the server keep retrying for a long time. Like maybe
```
Not on a liveness failure. (Those don't have a hope of fixing themselves on retry.)

In that case the server should probably then just forget about that `push_account_id`​, so that the client retries from scratch.

prakhar1144 replied:
See: https://github.com/zulip/zulip-architecture/pull/2#discussion_r2046582369

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,182 @@
+  
+- The server gives the client a list of the user’s currently valid `push_account_id`s
+  as part of `register` response, and keeps it updated through events.
+  - This allows the client to determine when it’s necessary to reregister the push device.
```
Copying over a comment from the old doc:
> I think the idea here (which should be spelled out) is that if the client doesn’t see on the list the `push_account_id` that it has for this account, then it knows it needs to reregister. The consequence of that is that the list should include all the `push_account_id`s that the client doesn’t need to act on by reregistering, which in particular means it should include any that the server is still working on setting up with the bouncer.
>
> [Then in addition to that information, i]ncluding the status (i.e. “has successfully registered with bouncer”, i.e. “has a non-null `device_id`") may be useful for a troubleshooting UI.

prakhar1144 replied:
Rewrote this section. Give it a read.

gnprice commented:
I shortened the title because "proposal" is redundant in the name of a ZA*P* :slightly_smiling_face:

prakhar1144 commented on zaps/0002-encrypt-push-notifications.md:
````diff
@@ -0,0 +1,236 @@
+- On the push bouncer, add a new table (parallel to `RemotePushDeviceToken`):
+  ```
+  class zilencer.models.RemotePushDevice:
+    device_id: RandomID[32]  # Unique key for this record, generated by bouncer.
````
I think, we don't need `device_id` to be random ?

gnprice replied:
I agree, this can just be a sequence number.

prakhar1144 replied:
```diff
-    device_id: RandomID[32]  # Unique key for this record, generated by bouncer.
+    device_id: int  # 64-bit integer primary key assigned by bouncer; not random.
```

prakhar1144 commented:
Addressed the review comments.

Posted 2 questions:
* https://github.com/zulip/zulip-architecture/pull/2#discussion_r2046898755
* https://github.com/zulip/zulip-architecture/pull/2#discussion_r2046776645

prakhar1144 commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,231 @@
+     doesn’t need to deal with retries here. Handling of 500 is discussed below.
+
+1. The server makes an API call to the push bouncer (authenticated with the
+   server’s API key) with arguments `user_id`, `realm_uuid`, `push_account_id`, and `encrypted_push_registration`.
```
I think we don't need the `user_id` here.

gnprice replied:
Yeah, agreed, that seems wrong.

prakhar1144 replied:
Removed it.

gnprice commented:
I have a workflow suggestion for this repo:

Instead of using the workflow we use in Zulip's software repos, I'd like to use a squash-merge workflow. That means:
* The PR branch isn't a draft of the sequence of commits to be pushed to the repo. Instead it's a draft of the desired total diff.
* When revising the PR, push the revisions as new commits on top, without force-pushing.
* When it's ready to merge, we'll use GitHub's "Squash and merge" button to squash it to a single commit.

This way, the commits in the PR are a record of its history during review. We know the end result will be a single commit — so, unlike with Zulip's main repos, there's no need to use the commit structure within the PR branch to express the details of the desired commit structure to be merged.

And this way everyone has a lot more options for looking at the PR after some revisions have happened and seeing exactly what's changed since the last time one looked, or exactly what the PR looked like at the time of a given review discussion.

(Cf. previous discussion of the trade-offs of squash-merge vs. Zulip's normal rebase-merge workflow, in the context of our main repos: [#git help > squash-merges @ 💬](https://chat.zulip.org/#narrow/channel/44-git-help/topic/squash-merges/near/2054898).)

prakhar1144 commented on zaps/0002-encrypt-push-notifications.md:
````diff
@@ -0,0 +1,231 @@
+  ```
+
+- We probably also need a `zerver.models.push_notifications.PushDevice`
+  for Zulip Cloud use with no third-party bouncer? Or we could make Cloud use a bouncer.
````
> We probably also need a `zerver.models.push_notifications.PushDevice`

I think, Tim meant a model parallel to `zilencer.models.RemotePushDevice` which will be present in zerver. 

`zerver.models.push_notifications.PushDevice` will always be present to store `push_public_key` etc.

---

 Rest of the discussion makes sense.

prakhar1144 replied:
Update: We don't need a new table for Zulip Cloud case.

See: https://github.com/zulip/zulip-architecture/pull/2#discussion_r2057009326

prakhar1144 commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -187,6 +201,12 @@ bouncer (or Google/Apple's servers, for that matter).
+  that have been stale for a week (use `last_accessed` to get the list).
+  The bouncer responds with the subset of those IDs for which either an
+  `expired_time` is set or no corresponding `RemotePushDevice` record exists.
+  Remove these devices from `PushDevice`.
```
> Greg Price:  
> ​​To unpack this a bit: I think the point is that when APNs/FCM tell the bouncer that a token was invalid or expired, the bouncer may need an asynchronous way to pass that on to the Zulip server.
​​
​​Whereas the spec above says that it’ll pass that on in the response to the server’s request trying to send the notification.
​​
​​I agree with “probably”.

> Greg Price
​​Looking back at our whiteboard diagram, I think an async path (with a request from the bouncer to the server to say a particular device_id​ is expired) is what we had in mind there, too.

> Timothy Abbott
​​From a network architecture standpoint, I don’t think we can expect the bouncer to be able to make any HTTP requests to the server — firewalls will prevent that. But we could make the server poll the bouncer on some regular basis to ask about expired devices or something.

gnprice replied:
For context, the comments quoted above were on an item in the original Paper doc which read:
> Does the error path from APNs/FCM to the bouncer to the server need to become asynchronous? Probably.

I think the current proposal's text here resolves that item and the comments on it: it's an async path for those errors to get from the bouncer to the server. (An item above this one turns those errors into either setting `expired_time` on the RemotePushDevice record, or removing that record, so that either way they become part of the bouncer's next response to this request.)

prakhar1144 commented:
New changes made are in a separate commit as discussed. See commit message for a summary.

PIG208 commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,231 @@
+    
+   **On error**:
+   - If it's a liveness error, don't retry and forget the `push_account_id`.
+     Pass the error to the server admin and the user.
```
Not sure if there are many examples of existing workflows for notifying admins. One might be having the notification bot send a message, like we do with channel subscriptions/bot errors.

timabbott replied:
I don' t understand the question -- why would administrators be involved here? This is, for now, just syncing a data set for an individual user of their registered devices.

prakhar1144 replied:
From the doc (Marked that comment unresolved):
> Greg said: Yeah, good question. If the server is persistently failing to reach the bouncer to register a device token, then I think the most important next thing to happen is to pass that error to the server admin so they can act on it. If they wanted push notifications and they aren’t actually working for some users, that’s a serious problem.

gnprice replied:
For reporting this error to the server admins, I'm imagining the same set of ways we use for other server errors. So it goes in the error log on the server, and I believe also an email, and whatever else.

prakhar1144 replied:
> The error is logged on the server and is also sent via email to the admin.

PIG208 commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,251 @@
+   and `push_public_key`.
+
+   The server creates a `PushDevice` instance. It replies to the client with a
+   success response, and asynchronously initiates the next step.
```
While the 500s from the bouncer service do not require the client to retry, what happens if it's the server that's getting 500s? Perhaps the client can keep using the same `push_account_id` until it is successful.

timabbott replied:
Possibly clients should treat 500s as a transient error that they should retry the same as a network error?

gnprice replied:
Yeah, the client will keep retrying with the same `push_account_id` until it receives a success response.

I'll post an edit to steps 2 and 3 to help clarify this.

gnprice replied:
(That edit: https://github.com/zulip/zulip-architecture/pull/2#discussion_r2057020659 .)

prakhar1144 commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,251 @@
+    token_kind: "apns" | "fcm"
+    token: str  # Provided by client.
+    push_account_id: int  # 64-bit random integer generated by client; need not be unique.
+    expired_time: datetime | None
```
Yeah, first point of Miscellaneous section mentions this.

prakhar1144 commented on zaps/0002-encrypt-push-notifications.md:
````diff
@@ -0,0 +1,251 @@
+  ```
+  class zerver.models.push_notifications.PushDevice:
+    user: ForeignKey[UserProfile]
+    encrypted_push_registration: Ciphertext  # A PushRegistration, encrypted with bouncer_key.
````
Makes sense. Moved it earlier.

prakhar1144 commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,251 @@
+   and `encrypted_content`.
+
+1. The client uses `push_account_id` to look up the right `push_public_key`,
+   decrypts `encrypted_content` with it, and ends up with the notification
```
Thanks for catching that, fixed it :)

timabbott commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,251 @@
+- Zulip Cloud doesn't use a third-party bouncer. To keep things conceptually
+  similar to the self-hosted case with the fewest different scenarios to think
+  through (reducing the risk of bugs), add a new table on the server (parallel
+  to `RemotePushDevice` added in the previous step).
```
These new tables for E2E encrypted notifications don't have the same format as the older tables used for push notification registrations history, like `RemotePushDevice`. All I was trying to say with the "parallel to" language is that I expect we'll want to keep the existing tables as-is (maybe sharing a base class), and plan to delete them in a few years once all self-hosted servers have upgraded, but that we're adding new tables, not extending the existing ones to have new fields.

We could consider whether that's actually the right decision -- but that can realistically be a thing to look at in engineering once we've mostly settled on a design. I currently expect it'll be less messy to just add new tables.

prakhar1144 replied:
Update: We don't need a new table (with a schema similar to `RemotePushDevice`, as proposed above) for Zulip Cloud.

See: https://github.com/zulip/zulip-architecture/pull/2#discussion_r2057009326

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -68,6 +69,15 @@ bouncer (or Google/Apple's servers, for that matter).
 
+- Make sure to correctly maintain the foreign key from `PushDevice`​ to
+  `UserProfile` during export/import. If not done carefully, it can leak
+  notifications to the wrong recipients.
```
This isn't exactly wrong but I think it overstates the danger. This table is like lots and lots of other tables in the Zulip data model in that it has a foreign key to other tables, and that if those foreign keys got corrupted in export/import then all kinds of bad things could happen, including sending data to the wrong users. I believe our export/import system is pretty good in general about getting that right, and isn't prone to that kind of bug. If that weren't true, that'd be a problem quite apart from anything this design is doing.

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -187,6 +201,12 @@ bouncer (or Google/Apple's servers, for that matter).
   attempts to tokens it already rejected.)
   - Likely a first version might just have a TODO comment on this point.
 
+- Configure a cron job that queries the bouncer with a list of `device_ids`
+  that have been stale for a week (use `last_accessed` to get the list).
+  The bouncer responds with the subset of those IDs for which either an
+  `expired_time` is set or no corresponding `RemotePushDevice` record exists.
+  Remove these devices from `PushDevice`.
```
If I'm following correctly, I don't think this `last_accessed` logic works.

This means the server would only learn that FCM or APNs reported a device ID as expired or invalid after the server had already stopped trying to send notifications, or had a pause of a week in doing so.

But what would cause the server to stop trying to send them? As far as the server knows, these are devices belonging to this user and so it should send notifications to them whenever the user wanted a notification, e.g. whenever they get a DM.

gnprice replied:
Instead:
* When there's an `expired_time` set, the bouncer doesn't need a list from the server of dubious devices — upon a check-for-errors request like this, it can reply with all the devices for that server that have `expired_time` non-null.
* If there's no RemotePushDevice at all on the bouncer for a given PushDevice on the server, then the server will learn that from an error if it ever tries to send a notification there in the future.

I guess a function of this `last_accessed` logic could be that it serves as a GC backstop: if a user stops getting DMs etc., and then their device quietly disappears and the bouncer learns that and removes the RemotePushDevice record, then a poll request based on `last_accessed` is a way for the server to learn it should remove the PushDevice record.

I'm not sure that does very much, though. IIRC, the only way APNs or FCM tell the bouncer that a given device has expired is if it tries to send a notification to that device. So if the user has entirely stopped getting notification-causing messages, then the RemotePushDevice and PushDevice records will stick around forever anyway (unless the user themself is deactivated, perhaps). And if they're getting even a trickle of them, then even without a `last_accessed`-based sweep, the server will learn it should remove the PushDevice on the second time it attempts to send a notification after APNs/FCM decide the device is expired.

prakhar1144 replied:
> This means the server would only learn that FCM or APNs reported a device ID as expired or invalid after the server had already stopped trying to send notifications

First point of **Miscellaneous section**:
> If APNs or FCM indicates to the push bouncer that the token is invalid
  or expired, the push bouncer sets `expired_time` to now. In that case
  — or if the push bouncer gets a notification for an invalid device or
  a device with `expired_time` already set — it returns an error to the server
  that the `device_id` is no longer valid, and the server removes it.
  
So, if server is still sending notification to such clients, cleanup would be immediate.

> When there's an expired_time set, the bouncer doesn't need a list from the server of dubious devices — upon a check-for-errors request like this, it can reply with all the devices for that server that have expired_time non-null.

I thought about this while writing, but I was aiming to cover the case where `RemotePushDevice` is not present together.

> So if the user has entirely stopped getting notification-causing messages, then the RemotePushDevice and PushDevice records will stick around forever anyway (unless the user themself is deactivated, perhaps).

This is helpful actually. Users can turnoff push notifications for a long time, even when they are using the app actively. So, FCM is not going to mark it expired and we'd be able to send them notification whenever they turn on.

For example, I have both RN and flutter app installed and push notification disabled for RN. I keep using RN app sometimes.

prakhar1144 replied:
> I guess a function of this last_accessed logic could be that it serves as a GC backstop

Yeah, this is doing GC. Not the primary async way to remove those entries.

gnprice replied:
> Users can turnoff push notifications for a long time […]
>
> For example, I have both RN and flutter app installed and push notification disabled for RN.

It sounds like you're talking about the settings on your device for whether the app can show notifications in the UI.

Those settings have no effect on the server sending notification-messages to the device — the server has no way of even knowing about them. They also have no effect on the notification-messages getting delivered by FCM or APNs to the app on the device. They only affect the step where, after the app has received the notification-message it got from the server via FCM or APNs, it decides to try to show you a notification in the UI: if you've disabled notifications in those settings, then that final step has no effect.

prakhar1144 replied:
One can turn-off [mobile-notifications](https://chat.zulip.org/help/mobile-notifications#mobile-notifications) for an org. 

>  I have both RN and flutter app installed and push notification disabled for RN.

Consider, logged in as different users (test & real account)

gnprice replied:
OK — but in that case I don't see what it has to do with which version of the app you're using, or with having both apps installed on a given device.

prakhar1144 replied:
Just a practical example for "Users can turnoff push notifications for a long time, even when they are using the app actively."

gnprice replied:
OK.

Back to the paragraph this subthread is about: it seems like this `last_accessed`-based sweep will have very little effect and we can do just as well without it. From above, after a user's device quietly disappears (because e.g. they uninstall the app or discard the device):

> […] if the user has entirely stopped getting notification-causing messages, then the RemotePushDevice and PushDevice records will stick around forever anyway (unless the user themself is deactivated, perhaps). And if they're getting even a trickle of them, then even without a `last_accessed`-based sweep, the server will learn it should remove the PushDevice on the second time it attempts to send a notification after APNs/FCM decide the device is expired.

The only thing this sweep changes is that if, after the device disappears, the user gets one notification-causing message but never gets a second one, then the server gets to delete the PushDevice record. That makes the case of one solitary such message turn out the same as the case of two or more of them, instead of turning out the same as the case of zero of them.

I think a Zulip server's database can afford the disk space of having the one-later-message-ever case keep the PushDevice record around just like the zero-later-messages case does.

prakhar1144 replied:
Alright, removed `last_accessed`. Made it simply:
> - Configure a cron job that queries the bouncer to get a list of device IDs
  from the `RemotePushDevice` table where `expired_time` is set, and remove
  these devices from `PushDevice`.

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -174,6 +187,7 @@ bouncer (or Google/Apple's servers, for that matter).
   - This allows the client to determine if the `push_account_id` it has
     locally for the account is missing on the `push_account_ids` map.
     If missing, it knows it needs to reregister the push device.
+    It helps to show an error to the user in case of liveness error.
```
I don't understand the last sentence here. What's a "liveness error" in this context? Is this saying that something helps for showing an error, or that showing an error is helpful?

prakhar1144 replied:
It's related to the error bouncer raises during registration:
>    **On error**:
>    - If it's a liveness error, don't retry and forget the `push_account_id`.
     Pass the error to the server admin. The user is notified about this
     error on the client using the `push_account_ids` map provided in the
     [register-queue](https://zulip.com/api/register-queue) response. (Refer
     to the [Miscellaneous](#miscellaneous) section for more details.)

gnprice replied:
Hmm, I still don't understand what this is saying. Can you clarify this part?

> Is this saying that something helps for showing an error, or that showing an error is helpful?

And if the former: what is the thing that helps for showing an error?

prakhar1144 replied:
This is addressing https://github.com/zulip/zulip-architecture/pull/2#discussion_r2050112579

> Is this saying that something helps for showing an error

Yes.

>  if the former: what is the thing that helps for showing an error?

`push_account_ids` map (which has `push_account_id` and `status`) will help the client.

gnprice replied:
Great, so this sentence could say:
> This data helps the client to show an error to the user in case of a liveness error.

I understand now what the sentence is saying. Next is the how and the why: how does the client use this data for that purpose, and why is that something we want to do? Why does the client care that there was specifically a liveness error?

For giving the user feedback that notifications aren't successfully set up, I think we probably want just a timeout: the client started trying to tell the server about its token more than, like, a few minutes ago, and the status is still that it hasn't succeeded yet.

prakhar1144 replied:
PUSH_REGISTRATION_LIVENESS_TIMEOUT is 24 hours. 

So, are you suggesting:
* Storing the timestamp in client when the registration request was made and server responded with 200.
* Then, in future, when the user reopens the map and the `push_account_id` is missing in the map + 24 hours passed; we can show the error and retry.

gnprice replied:
No, I meant what I said.

If the user installed the Zulip app and logged into Zulip 5 minutes ago, and they're not getting notifications when people DM them, they're going to be wondering what's up because "the mobile app" clearly isn't working. So the app should acknowledge that.

At that point the server will still be periodically attempting to retry the registration with the bouncer (because PUSH_REGISTRATION_LIVENESS_TIMEOUT hasn't expired). In other words it will be still attempting to recover from the outage. But that doesn't mean there isn't an outage.

prakhar1144 replied:
Okay, simply client side handling.

prakhar1144 replied:
Updated in the 7th commit.

gnprice commented on zaps/0002-encrypt-push-notifications.md:
````diff
@@ -48,11 +48,16 @@ bouncer (or Google/Apple's servers, for that matter).
   ```
+
+- The push bouncer has an asymmetric key `bouncer_key`, of which the
+  public part is baked into clients.
````
It occurs to me we should have a plan to be able to rotate these: make a fresh key, then distribute clients that can work with both the old and the new key, then at some point later stop using the old key.

That rotation can't be done in a hurry without disruption, because many users' devices are slow to upgrade the app — like many months to get past 90% upgraded. (Though that's probably a lot more than 90% of new logins, because I'd expect the bulk of logins happen within hours of first installing the app.) But it'd be prudent to be able to do it occasionally, even if slowly.

prakhar1144 replied:
[#api design > E2EE push notifications @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/E2EE.20push.20notifications/near/2163334)

prakhar1144 replied:
Plan included in the 8th commit.

gnprice commented on zaps/0002-encrypt-push-notifications.md:
````diff
@@ -0,0 +1,263 @@
+    expired_time: datetime | None
+  ```
+
+- Zulip Cloud doesn't use a third-party bouncer. To keep things conceptually
+  similar to the self-hosted case with the fewest different scenarios to think
+  through (reducing the risk of bugs), add a new table on the server (schema
+  similar to `RemotePushDevice` added above).
````
There should not be any additional table for Zulip Cloud. That'd be an example of making things work differently from the self-hosted case in a way that risks bugs (plus is more work).

Instead, in Zulip Cloud, the server should have a PushDevice model just like any other server; and the bouncer should use RemotePushDevice records for Zulip Cloud users' devices the same way it would for any other server.

The absolute simplest design, conceptually, would be that Zulip Cloud as the server makes an HTTP request to the bouncer the same way as any other server. But a background fact about the push bouncer is that it's the same web server and database as Zulip Cloud, just an additional set of endpoints which it responds to. So, as an optimization, we can have it skip the HTTP request and instead invoke the bouncer functionality through Python function calls within the server codebase. I believe that design is the status quo.

That optimization makes the design more complicated to think about: there are two different ways the bouncer functionality gets called, and two different ways the server can relate to the bouncer.

The point here is that we shouldn't compound that complication by having the database look any different in Zulip Cloud from how it looks in the self-hosted case, if you were to look at the bouncer's database tables and the server's database tables together.

prakhar1144 replied:
That's correct... but,

When a self-hosted installation decides to NOT use the push bouncer, the above approach won't work, right ?
They don't have `zilencer` (bouncer) tables.

We can do what you said for Zulip Cloud, but we can think Zulip Cloud as the case of a isolated server which decides to NOT use push bouncer (hence any of it's table) -- hence covers the above mentioned self-hosted case.

Or, Zulip Cloud works as you explained & we can document the "self-hosted server not using bouncer" case separately. I understand this case is rare but we'll need this for obvious reasons.

gnprice replied:
If a self-hosted server wants to send notifications but without using [Zulip's mobile push notification service](https://zulip.readthedocs.io/en/latest/production/mobile-push-notifications.html), i.e. the bouncer that runs in Zulip Cloud, then it should run that the same way as Zulip Cloud does: by using itself as the bouncer.

We shouldn't have any code (beyond perhaps a bit of settings code), and definitely shouldn't have any different way of arranging tables, that's specific to a server sending notifications without a separate bouncer and yet also in a different way than Zulip Cloud does so. That way lies bugs, and also an extra burden for us in writing that code, reviewing it, and then for every future change thinking through how it interacts with that case.

prakhar1144 replied:
Updated to align with the discussion.

gnprice commented on zaps/0002-encrypt-push-notifications.md:
````diff
@@ -0,0 +1,263 @@
+  ```
+  class zerver.models.push_notifications.PushDevice:
+    user: ForeignKey[UserProfile]
+    encrypted_push_registration: Ciphertext  # A PushRegistration, encrypted with bouncer_public_key.
````
Here's an adjustment that wasn't in the original draft:
```suggestion
    encrypted_push_registration: Ciphertext | None # A PushRegistration, encrypted with bouncer_public_key. Null once bouncer has confirmed the registration.
```

This field is never gets used once `bouncer_device_id` is set to non-null. So it can be cleared out at that point.

More conceptually: this field basically describes a pending request that needs to be sent to the bouncer. Once that request has been successfully made, and the acknowledgement recorded in the form of `bouncer_device_id`, that information is obsolete.

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,263 @@
+   (Note: We’ll likely want settings to require E2E encryption; that can come
+   after the feature is complete).
+
+1. The client generates a random 64-bit integer `push_account_id`.
+   It builds a JSON object `push_registration` of type `PushRegistration`,
+   and encrypts it to `bouncer_public_key` as `encrypted_push_registration`.
+
+   It also generates an asymmetric key pair (`push_public_key` and
+   `push_private key`).
+
+1. The client makes an API call to the server (authenticated with the user’s
+   API key) with arguments `push_account_id`, `encrypted_push_registration`,
+   `push_public_key`, and `token_kind`.
```
Edits in particular to help clarify what happens if a previous request from the client didn't get a successful response from the server (following up on https://github.com/zulip/zulip-architecture/pull/2#discussion_r2052975643 ):

```suggestion
1. If the client already has `push_account_id`, `push_public_key`, and
   `push_private key` set on the `Account` record for this account,
   it proceeds to the next step.

   Otherwise, the client generates a random 64-bit integer `push_account_id`
   and an asymmetric key pair `push_public_key` and `push_private key`,
   and stores these on the `Account` record.

1. The client builds a JSON object `push_registration` of type `PushRegistration`,
   using the current time for `timestamp`.
   It encrypts this to `bouncer_public_key` as `encrypted_push_registration`.

   The client then makes an API call to the server (authenticated with the user’s
   API key) with arguments `push_account_id`, `encrypted_push_registration`,
   `push_public_key`, and `token_kind`.
```

The specification for `timestamp` I think also wasn't written down before.

gnprice commented on zaps/0002-encrypt-push-notifications.md:
```diff
@@ -0,0 +1,263 @@
+    …
+    push_account_id: int  # 64-bit random integer generated by client; unique here.
+    last_acked_push_token: str  # The last push token the server acknowledged receipt of from this client.
+    push_private_key: PrivateKey  # Private part of the asymmetric key pair generated by the client.
```
Noticed this in the context of https://github.com/zulip/zulip-architecture/pull/2#discussion_r2057020659 :
```suggestion
    push_private_key: PrivateKey  # Private part of the asymmetric key pair generated by the client.
    push_public_key: PublicKey  # Public key corresponding to push_private_key
```

The client needs to remember the public key at least long enough to retry the registration request.

prakhar1144 commented:
Addressed the last batch of comments. I'll post a design for https://github.com/zulip/zulip-architecture/pull/2#discussion_r2056977181 in CZO.

timabbott commented:
Merged via #3. I verified all the comment threads above are resolved in the merged draft.
